### ISSUE TITLE
Unsaved buffers aren't restored if there is no project folder added to the window

### ISSUE THREAD

**Issue by Nek80** [atom/atom#10474]
According to this [release notes for 1.4.0beta](https://github.com/atom/atom/releases/tag/v1.4.0-beta0) and this [issue report](https://github.com/atom/atom/issues/942) Atom is supposed to automatically save any unsaved/untitled buffer upon closing.

Steps to reproduce:

1. Open Atom.
2. Open a new tab
3. write something without saving it
4. close atom (click X or file->exit)
5. reopen atom
6. the unsaved new untitled tab is missing. It should be there with all the content that you wrote.

This is important feature for me, please fix it. Thanks.

**Comment by 50Wliu** [comment]
I can't reproduce this.  Can you please try `atom --safe`, and exactly what Atom version are you using?  Thanks.

**Comment by mvisat** [comment]
I have the same issue. I'm using Atom 1.4.0. Atom doesn't save untitled buffer if there is no project folder opened. But if I add any project folder, the untitled buffer is saved. Thanks.

**Comment by mnquintana** [comment]
Thanks @mvisat! I can reproduce this now with these steps:

1. Open a new window. Make sure it doesn't have a project folder added.
1. Type some stuff in an untitled buffer
1. Quit Atom
1. Reopen Atom
1. Atom won't restore what you wrote in that untitled buffer

**Comment by GuacheSuede** [comment]
Is this related ? #10480

**Comment by mattpker** [comment]
It appears that they were aware of this issue in the pull request:

https://github.com/atom/atom/pull/9968#issuecomment-163872905

So this does not appear to be a bug, but more like an attended effect. I posted on that pull request why I think that this is an oversight and needs to be changed to something better. Ill repost that here as this is probably a better place for that.

--

This is a great new feature and I really love it, but I think more work needs to be done when you have not opened a project. A lot of the time when I need to jot down some notes I will open a new atom window (not apart of a project). When I go to close that window or restart my computer, the pop up to save the untitled notes was really great, because a lot of the time I would forget to save a note that I actually needed to keep. Because this was not a part of a project, I just lost all my notes. I think having a feature that only works on projects is a bit of an oversight, it gets you in the mindset that "It's okay if I close this without saving, it will be back when I re-open it." and then when it was not opened as a project it is gone with no prompt warning you of this.

I suggest adding the prompt back for untitled tabs that are not apart of a project, or making all non project tabs all be apart of some "no project" project or something. I hope that makes sense, let me know if you have any questions.

**Comment by Meligy** [comment]
I think the missing prompt + no save for non project files should be considered a bug.

The original pull request covers saving project files only, which is why the the side effect of losing the warning for unsaved files is a bug.

Of course I'm very grateful to see this feature come slowly to Atom.
After all, I don't pay Atom anything and am grateful for everything the contributors are giving away.

However, I was really hoping for this to have included unsaved files.
This is the only reason I keep going back to Sublime for draft/note writing in Markdown, although Atom has much better Markdown preview.

**Comment by soomtong** [comment]
Hi. I added screenshot for `more-information-need` tag

I have a report for a situation about untitled document condition
it is related https://github.com/atom/atom/issues/10540 https://github.com/atom/atom/issues/10474
but still produced like below 

- OSX 10.11.2
- Atom 1.4.1

## 1. With Project
  ![image](https://cloud.githubusercontent.com/assets/906974/12618116/e1c763b4-c555-11e5-83e9-74e44860c785.png)
  re-run atom, no problems here

## 2. Without Project, just one untitled document
  ![image](https://cloud.githubusercontent.com/assets/906974/12618207/3fba9dd8-c556-11e5-80ba-218f17c216f8.png)
  re-run atom, disappeared latest text
  ![image](https://cloud.githubusercontent.com/assets/906974/12618246/5e4bc8d0-c556-11e5-8349-96b88908ab82.png)
 
i think that maybe it's a base directory setting related. i did check and update `Project Home` value
same problem here.

i guess, atom didn't figure out it's default temporary document location.

Thanks Atom!

**Comment by Ashkaan** [comment]
Anyone interested in making a package until it's eventually added to core functionality?

**Comment by curtiszimmerman** [comment]
None of this works for me at all. I am trying to evaluate Atom (again, for the third time) to see if I can yet ditch ST3 in favor of Atom, and this feature is non-negotiable for me. I take my laptop to and from work and being able to restart my session immediately from wherever I left off in the way that ST3 does this is a requirement for me to use any other editor. I have tried all the versions from 1.4.0 to 1.5.3 to 1.6.0-beta6, using precise instructions from the issues here on GH and also from someone using the same OS as I (LM17.3 Cinnamon) in #atom on Freenode.

I desire that I can just fire-and-forget with the close button, and Atom will return to me not only my open tabs of named files, but also unsaved changes in named files, as well as unsaved changes in unsaved ("untitled") files. I want to be able to do this regardless of how and where I opened Atom, and I do not want my unsaved files to be tied to a specific project. I want unsaved files to be restored no matter which "project"/directory/location I start Atom from.

So this doesn't work. I went looking for help and followed these directions to a tee, which did not work for me: Open Atom from within some "project"/directory. I did so via Guake's terminal with `/usr/bin/atom-beta ~/src/project` and opened a new file, then typed some crap into it ("foo"). I then closed the editor with `Alt+F4`. Then I reopened the editor using the exact same command in the same Guake terminal I used to open it previously. I am then greeted with a blank pane.

I have enabled the core "autosave" package. I can't use Atom until I am able to restore untitled/unsaved data from any location, and even though I can't get this to work, I think it is a bad idea to tie this behavior to "projects" instead of just making the unsaved data generally available. Why should I have to remember that my scratch pad / source code snippet was created with project/directory X instead of just having the computer do the work for me and remember my session? Why do I need to remember anything? That's what I'm using the untitled tab for in the first place.

**Comment by curtiszimmerman** [comment]
Ah, nope. Turns out it doesn't persist across reboots. For some reason, the minute I reboot my machine, I am no longer able to restore from my sessions like I was able to before. I can see that Atom (both the core package autosave and the installable save-session) appear to write text buffers to files in their respective locations in ~/.atom/, and those buffers still are around when I restart. But it doesn't matter which of those I have installed (or both or neither), when I start Atom -- from the LM quick launch bar which starts `/usr/bin/atom-beta %U`, from the command line with `/usr/bin/atom-beta ./`, or using `Alt+F2` and typing `/usr/bin/atom-beta` -- it brings up a very old text buffer, from before I got this to work by uninstalling everything and hosing `~/.atom/`. Whatever the case, Atom isn't saving my session, or the functionality is broken, or something. The "save buffer" is getting overwritten on startup.

**Comment by h-collector** [comment]
Atom 1.6.0, 1.6.1, Linux Mint 17.3
I had almost the same problem.
Unsaved changes, untitled document, closed/opened tabs weren't preserved between sessions.

But in my case I narrowed the culprit to atom-community/formatter package.
I had installed: formatter-coffeescript, formatter-coffeescript, formatter-coffeescript, formatter-coffeescript, enabling main package or any other depending on formatter service was breaking session saving :/

**Comment by elad** [comment]
Bit me with version 1.6.2 on Mac OS X.

This is very frustrating. It's about the only important feature for me in a text editor. I honestly don't understand why this can't just work out of the box. I also don't understand why the concept of "projects" even exists for a text editor.

*sigh* Back to Sublime.

**Comment by Ashkaan** [comment]
@elad Waaayyy ahead of you.. I went back to Sublime after I realized that no one is interested in what I also consider to be the most important feature.

**Comment by sdaddy** [comment]
I can't believe this is still a problem. What editor doesn't ask to save a file when you close it? Seems this has fallen under "We don't use it like that so it's not a priority" types for the developers. Sheesh.

**Comment by ArniPL** [comment]
@sdaddy Atom does ask to save untitled buffers and changed files but **only** when you try to close it's tab.

But yeah, it does not ask when you close entire editor. And then it doesn't restore changes when you have no project folder added.

**Comment by Meligy** [comment]
> But yeah, it does not ask when you close entire editor. And then it doesn't restore changes when you have no project folder added.

Exactly. And that's a big problem compared to Sublime. In Sublime I can just keep having drafts, kill it any time I'm presenting or having a sole-use of my laptop, and open it later with all drafts intact. The same even if my laptop crashed while Sublime was opened.

**Comment by elad** [comment]
I don't mean to sound negative, but Atom is DOA. It's ridiculously slow compared to Sublime and it lacks features that I consider fundamental. I'm following [the same issue on VS Code](https://github.com/Microsoft/vscode/issues/101).

**Comment by JuDelCo** [comment]
Sadly, i'm surprised no editor (well... sublime "alternative") has this core feature properly implemented yet, but i'll wait ...

**Comment by Ashkaan** [comment]
From my experience, ATOM can be made to look pretty, but it doesn't
function anywhere near as well as Sublime.. nothing does.

On Fri, May 13, 2016 at 3:36 PM JuDelCo <notifications@github.com> wrote:

> Sadly, i'm surprised no editor (well... sublime "alternative") has this
> core feature properly implemented yet, but i'll wait ...
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/atom/issues/10474#issuecomment-219175578>
>
-- 
Ashkaan Hassan, Esq.
Attorney-At-Law
*Ashkaan.Info <http://www.ashkaan.info/>*

CONFIDENTIALITY NOTICE: The information contained in this electronic
message and any attachments may contain confidential and privileged
information.  If you are not the intended recipient, please notify the
sender immediately by return e-mail, delete this e-mail, and destroy any
copies.  Any dissemination or use of this information by a person other
than the intended recipient is unauthorized and may be illegal.  Unintended
transmission shall not constitute waiver of any applicable legal protection
afforded to this electronic message and any attached documents.

**Comment by curtiszimmerman** [comment]
Being able to drop whatever I'm doing, shut down my machine, move somewhere, and start it back up exactly where I left off is crucial. It's the only reason I don't stick with Atom when I try it every six months or so. Sublime does this perfectly, and I can't consider any replacement that doesn't do this flawlessly, and I judge very harshly on this one point. Sessions must be managed/saved/restored totally without a hitch every time, or I won't move on from Sublime. Atom has improved hugely since its inception, and I'm sure one day it will beat anything else out there. I really want Atom to succeed, but this issue makes it a non-starter for me.

**Comment by splashx** [comment]
Unfortunately I didn't come across this issue until now - sadly I've lost important work because Atom doesn't implement this *no-brainer* function.

**Comment by curtiszimmerman** [comment]
@splashx So I *think* this problem has been solved at least at some point before 1.8.0-beta. But it definitely wasn't fixed in 1.4.0 like they said in that link, since I couldn't get it to work right in 1.6.0-beta. I'm using 1.8.0-beta3 and it is preserving my workspace across reboots, including unsaved file buffers. I have performed a hard power cycle (to try simulating a crash) and it still preserves my workspace, including unsaved file buffers. Would you mind putting 1.8.0-beta through some testing to see if it works for you?

**Comment by Meligy** [comment]
@curtiszimmerman I just tried now, Atom 1.8-beta3 on Mac OSX 10.11.5.
It failed.

Steps to reproduce:

1. Open Atom
2. Go to side menu, right click every root-level folder and remove project directory
3. Create new Atom window and close the old window
4. Write something in the default tab in this window, or add more tabs and write in them
5. Quit Atom
6. Open Atom, your tabs are gone

The reason for the steps is that Atom will save your data when there's a project folder. It also wouldn't keep your last removed project folder as removed when you open.

So, even if you remove a project folder, when you quit and re-open Atom, you'll find the folder added, and you'll (most likely) find your unsaved tabs.

But if you haven't added a project folder (a folder in the sidebar), it does not remember your tabs.

The inconsistency makes it easy to trust Atom to save the tabs, and accidently get into a situation when it doesn't save them.

**Comment by curtiszimmerman** [comment]
1.8.0-beta3 does not work on Linux, either, when adding the steps of removing project folders. That is totally unacceptable and too bad! It totally has linked saved sessions to this awful "project folders" idea. Unacceptable. Back to ST3.

**Comment by splashx** [comment]
@Meligy got it right. So right that I actually found my "lost" work a couple of minutes later after my post - I just couldn't believe the data was not anywhere in the hard disk so I grep'ed `~/.atom` for some keywords I knew the `untitled` had. It turns out I found a match, but just because I accidentally had a project open on that window. I couldn't remember it because it was totally unrelated to the `untitled` tab. The grep showed me the link between the `untitled` and the project folder, so all it took was "atom /path/to/unrelated/project/folder" and the `untitled` I thought I had missed showed up.
But that was *pure* luck - I initially thought I didn't have a project set for that `untitled` tab, so I started to pull my hair and eventually ended up here on this thread. 

Bottom line: it's not fixed, so keep your eyes open to not start writing important stuff whitout having a project folder opened (e.g. note taking).

**Comment by designorant** [comment]
Just for the records, the above is still the case in `1.8.0`, and it is very confusing indeed.

Having seen the unsaved files preserved within my projects, I assumed this was Atom's default behaviour (as part of **Back Up Before Saving** feature) only to find the file I've been using for daily note taking has been lost after a restart prompted by an update.

**Comment by paxpacis** [comment]
Just here to confirm. Issue is still on Mac OSX Atom version 1.8.0 with and without a project dir btw.

**Comment by sgoendoer** [comment]
Aaaaand this is still an issue in Atom 1.8.0 (Mac OS X)

**Comment by soulmachine** [comment]
This issue still exists in Atom 1.9.0(Mac OS X)

**Comment by paxpacis** [comment]
I found sort of work-around.

If you start atom with cmd + space it wil work. but if you start atom from another directory or by clicking a file it does not work as expected. It looks like it keeps the unsaved files per starting point.

So if you start atom with ~/blah123/ and have unsaved files they wil be saved if you start atom again with ~/blah123/ if you start atom with ~/blah1234/ it wil not load the ~/blah123/ unsaved files..

ym2cents

**Comment by ctwise** [comment]
Starting Atom using a launcher (Spotlight or Alfred) doesn't keep the
unsaved buffers for me. Starting it w/o a path also doesn't keep the
unsaved buffers. The unsaved buffers are attached to paths.

**Comment by paxpacis** [comment]
@ctwise 
"The unsaved buffers are attached to paths."

That is what I think too..

**Comment by Meligy** [comment]
They are attached to folder paths because they only work in projects, and projects are attached to folders. See my explanation above in this same thread https://github.com/atom/atom/issues/10474#issuecomment-174706820.

**Comment by RSully** [comment]
Are these buffers saved to disk, just not opened? Where are they saved? I lost one and just spent a long time trying to figure out where the data might be, as it was quite important. I `ack`'d around ~/.atom, ~/Library, /var/folders, etc for substrings I could remember- no luck.

**Comment by sbowman** [comment]
If you're not going to automatically save the buffers, could you at least warn me that I have unsaved work?  1.9.6, and this is still an issue.  Just lost a ton of work.

**Comment by SnoringFrog** [comment]
@RSully That is what I would like to know. Now that I know this is an issue I can work around it until it's fixed, but right now I'm just trying to find what I lost, if at all possible. I restored `.atom/blob-store`, and `.atom/storage` to earlier versions and then tried reopening atom as I usually word after a reboot, but it still just loads what it thinks my current session is for some reason. I can't figure out how/where unsaved content is tied to the project/directory at all.

**Comment by Extazium** [comment]
There is a workaround, just install this package : https://atom.io/packages/save-session
It works perfectly fine for any unsaved tab. 
(Tested on Win8.1 with Atom 1.9.9)

But there's just one problem : you need to open Atom (with the `atom` command or shortcut) first before you open it in a specific directory. If you don't do that, you lose everything. Happened to me this morning and you guessed it, I'm kinda mad for losing my notes.

**Comment by biax** [comment]
npp, sublime all have properly working session save/restore. such an amazing feature. a 6 months issue and still not solved? not too impressed.

**Comment by fseesink** [comment]
And it's STILL there in Atom 1.10.1.  (I'm on Mac OS X 10.11.6.)

I would very much like to like this editor.  It has so much potential.  Being an open-source project and having such a nice package setup, there's that freedom to easily add features and hack on it for those so inclined.

And in fact, it seems this particular issue has been around so long that there WAS a package to deal with this:  https://atom.io/packages/save-session

But...
"Project Status: Deprecated
All of Save Session's functionality is now included in Atom core."

Great... except it's not.  Seriously?

I read up on **Brackets** and **Atom** over a year ago.  And at the time I thought, "Interesting.  But the reviews indicate they're a bit slow."  And when you work in an editor, the one thing you should never have to do is wait for your editor.  Your editor should wait for you.

Then recently I was reading up on various topics including Ansible, and I found that many folks are using Sublime as their text editor of choice, while others chose Atom as it's very Sublime-esque and they even had packages to do proper syntax highlighting of Ansible playbooks (which are just YAML files), but not just basic YAML but Ansible-specific keywords/etc.  That caught my eye.

And I went out and researched.  And I found out about all the Atom packages.  Wow.  And how it really seemed to have a decent following.

So I downloaded and installed it.  And I was so impressed that within a day I was using it regularly, and within the week I had installed it on all my regular machines and configured it as my default editor.  I haven't found it slow for my purposes at all.  And the package feature is just incredible.

But this one issue is a real deal breaker.  As others have mentioned, what editor today doesn't support this?  And not in some fragile, maybe-it-works-maybe-it-doesn't sorta way. (*)  But simply do not allow me to lose any of my work without me saying so.  When I quit the app with unsaved tabs/windows, regardless of how, warn me.  And if you're going to quit/close outright, then when you fire back up, you darn well better have my text there in whatever windows/tabs I had open when I quit.

**BareBones TextWrangler** (free) does this.  **Sublime** does this.  Most editors do this.  The fact Atom doesn't is shocking considering the rapid pace of development.  What this makes me wonder is whether the developers are focusing on "cutesy" features and not focusing on fundamentals.  If you don't handle the basics well, I honestly don't care about your bells and whistles.  I can't trust your software.  But do your fundamentals properly, and the rest will follow.

Please, GitHub, your _About_ tab says this program is "<> with {heart symbol} by GitHub".  How about showing this editor some love and fixing this fundamental feature?

----
(*) I'm experiencing this problem on my iMac but so far my MacBook Pro seems to save session data... and I cannot for the life of me figure out what's different between them.  But when it happened on the iMac, it wasn't a happy day.  Whereas in the past I could easily be taking some notes in TextWrangler, have 5-10 tabs open, nothing saved, and quit without ever worrying anything would be lost, I happened to hit Cmd-Q instead of Cmd-W (something that happens every now and again), and poof!  Work gone.  No warning dialog.  Nada.  I didn't think of it, believing it would simply restore things.  But no.

**Comment by JeremyIglehart** [comment]
This should be labeled as a bug, and be an urgent issue.  What gives?

I told my wife Atom is great! You should use it for your editing for x, y, z reasons.

Then I noticed (after installing atom again on a new computer) that atom wasn't saving my session data if I closed atom without saving - I would lose my work...  Lost some important notes and phone numbers and call logs from dealing with customer service people.  GRRR

Then today I hear my wife cursing under her breath, I said "What's going on, can I help?" And she was upset cause yesterday she spent a whole hour working on a document and forgot to save it, now she's lost her work.

Thanks Atom, for making me look like a jerk.

I don't understand why this is not labeled as a bug.  Is there an anger emote on Github?

**Comment by iolsen** [comment]
I definitely understand the frustration with this behavior. We should be either prompting to save these buffers or maintaining their state. I've reproduced the behavior and we'll look into this soon.

**Comment by JeremyIglehart** [comment]
Thank you @iolsen for responding to our plight.

Might I add my 2¢ with everyone else and say that Sublime Text has a default behavior I think has become a standard among text editors, and other things like google drive applications, gmail, etc.

If you're quitting, and you haven't saved, simply maintain the state till the next time you open.

If people don't want this and would rather have a prompt - make that a setting checkbox to switch behavior.

Personally, I am always a fan of [less prompts](https://en.wikipedia.org/wiki/Modal_window) in general - I don't want it to prompt me, I just want to quit and have it come back the way I quit, saved or not.

**Comment by amavlyanov** [comment]
The default behaviour of Atom is confusing people. This should be the no 1 priority to fix: Atom should keep all the opened buffers in all opened windows and restore them after accidental closes (like cmd+q) / crashes. And if such buffers contain previously saved files - those files should not be overwritten, since the changes might not be tracked by some VCS.

This should be done the way sublime text 2/3 do this. Exactly the same way. It's like what people expect from the 21st century program.

**Comment by Ashkaan** [comment]
Guys, this is pointless.  Clearly, they aren't trying to do what Sublime does.  This is a different product with a different purpose.  I suggest we close this thread and stop calling it a bug.  This is a really easy fix and if they intend to fix it, they would have.

**Comment by 50Wliu** [comment]
@Ashkaan See @iolsen's [comment](https://github.com/atom/atom/issues/10474#issuecomment-249301157) above.

**Comment by iolsen** [comment]
> @Ashkaan See @iolsen's comment above.

👍 

Sorry for the delay but this is next on my list. We have a release queued up for early next week that this won't make, but a fix should be in the following release.

**Comment by Ashkaan** [comment]
Oh wow.. ok!

**Comment by iolsen** [comment]
Big thanks to @stereobooster for fixing this! It'll be in 1.13.

**Comment by ArniPL** [comment]
Great that it's finally resolved in some way! 

But as I see it, it's still not perfect. Now Atom will prompt to save the unsaved buffers, but won't maintain editor state transparently. So it's still impossible to restore editor state without project folder. Is this intended behavior? Thanks!

**Comment by 50Wliu** [comment]
Yes, I agree that it would be nice to have the behavior that @ArniPL is asking for.

**Comment by smaier** [comment]
god, please fix this behaviour - such a great Editor

**Comment by biax** [comment]
cant believe this topix is  still open for nearly a year. quite disappointing honestly. instead of having a few editors for various things, i now simply open a few sublimetext windows. and the memory footprint on atom..... sigh.

**Comment by Ben3eeE** [comment]
This issue is not open. The fix for this issue has been merged and will be available starting with Atom 1.13.

**Comment by kstenerud** [comment]
Is it in 1.13 beta 5? I just tried and it doesn't preserve open buffers.

**Comment by Ben3eeE** [comment]
The change that closed this issue is to ask for confirmation to save unsaved buffers when closing a window without a project folder, not to preserve them.

**Comment by curtiszimmerman** [comment]
This is ridiculous. Screw it. I'm just giving up on Atom. Now there is a barrier to shutting this thing down. This is not even close to what we're talking about.

**Comment by amavlyanov** [comment]
@Ben3eeE it's not a fix, it's a workaround to not loose the data. When will the real fix will be available for this problem?!

**Comment by kstenerud** [comment]
Alright, I can see where this is going. I'll just switch back to sublime text.

**Comment by ctwise** [comment]
Please reopen this issue. The issue isn't about a warning, it's about saving buffers.

**Comment by iolsen** [comment]
My apologies for cavalierly closing this issue without realizing the signal that sent. Fixing the potential data loss was a priority and I'm happy that has been fixed, but I agree that preserving the buffer state in this scenario is the more elegant solution and should still be done. I've created #13318 to that effect.

**Comment by gmmedia** [comment]
Wow, after a year still not fixed.

I also give up on Atom. Back to Sublime.

Great work... NOT

**Comment by kstenerud** [comment]
That's not at all constructive. Give it a rest.

**Comment by ArniPL** [comment]
@gmmedia you can always submit a pull request with working solution if you need this fixed. Otherwise why do you complain about **free software** on a development forum? No one needs this here. Besides, this issue moved to #13318.

**Comment by Meligy** [comment]
If you need this feature in a Chrome based editor, you can use Visual Studio Code Insiders Edition, which started Rolling this Freddie already and actively working on it.

As for Atom, it's obviously not happening soon, and as mentioned above, no one had the right to blame free software, so, unless you have the skills to implement it yourself, Atom is a dead end for this scenario in the foreseeable time.

**Comment by 50Wliu** [comment]
I'm locking this issue as it's gotten out of hand.  The new issue is at #13318: please subscribe to that one instead.  Thanks!

**Comment by lee-dohm** [comment]
I've deleted a number of off-topic and solely negative comments here. Those that have had their comments deleted should refer to the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) and consider themselves officially warned. Let's keep everything professional and constructive here.
